src/main/java/uk/gov/gchq/palisade/util/ResourceBuilder.java

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright 2020 Crown Copyright
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.gchq.palisade.util;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import uk.gov.gchq.palisade.resource.LeafResource;
+import uk.gov.gchq.palisade.resource.ParentResource;
+import uk.gov.gchq.palisade.resource.Resource;
+import uk.gov.gchq.palisade.resource.impl.DirectoryResource;
+import uk.gov.gchq.palisade.resource.impl.FileResource;
+import uk.gov.gchq.palisade.resource.impl.SystemResource;
+import uk.gov.gchq.palisade.service.ConnectionDetail;
+
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Provides a common set of utilities for constructing resources with all parents
+ * automatically constructed recursively. This primarily targets filesystem-like
+ * resources (Files, Directories etc.)
+ * Internally, the resourceId is converted to a URI.
+ *
+ * Can produce any of the following output types:
+ * - {@link FileResource}
+ * - {@link DirectoryResource}
+ * - {@link SystemResource}
+ * Any parents automatically constructed will also be from this collection.
+ *
+ * If another method of creating a resource is required (i.e. directly using strings)
+ * there is no guarantee that this can correctly resolve parents. Instead use the
+ * methods provided by the appropriate resource impl.
+ */
+public class ResourceBuilder {

```

Comment:
This isn't a resource builder, it's a FileResourceBuilder and if it is to align with the static lambda builder patterns I have shown you all it should be embedded into the FileResource class. Bounce this off Ashley.

Reply:
I would disagree, this class works out what sort of resource to build including whether it needs to build up the parent hierarchy. At the moment this does look very much like it is just a file/directory resource builder. But if we was to expand the resource schemas supported to include databases or streaming topics then this central builder would allow the building of those from the same create from URI method. Therefore for now am happy to keep this here.
The question for me is whether it is right for us to be making those decisions based of the URI. In an ideal production use that information would come from a data catalogue which would tell us whether it is a FileResource or StreamResource and all the required attributes. However we are having to build scaffolding around Palisade to stub that kind of functionality and I see this as part of that scaffolding.
As for the static lambda builder pattern, I think that will come in later on for each of those File, Directory, System Resource classes in a different PR

Reply:
OK, I'm fine with that.